### PR TITLE
Enable runtime DSPARK target feature capture

### PR DESCRIPTION
**Pull request by tejesh-sarvam**
## Motivation

DSPARK requires auxiliary target hidden states from configured decoder layers. Some SGLang models expose native DSPARK or DFLASH capture hooks, but compatible models using the standard residual-stack and `LayerCommunicator` contracts do not.

Previously, supporting such a model required adding model-specific DSPARK code or selecting another compatible model implementation. This PR adds an executor-side fallback for compatible models while preserving existing model-native implementations.

Sarvam target verification must also honor `--speculative-attention-mode`. Because `TARGET_VERIFY` is distinct from ordinary decode, it previously selected the prefill attention path even when speculative attention was configured for decode.

## Modifications

- Add a runtime DSPARK auxiliary-hidden-state capture adapter for compatible residual-stack models.
- Prefer existing model-native `set_dspark_layers_to_capture` and `set_dflash_layers_to_capture` hooks.
- Use the runtime adapter only when DSPARK capture is requested and neither native hook exists.
- Validate the runtime capture contract before attaching the adapter:
  - explicit, ordered and unique target layer IDs;
  - standard `model.layers` decoder structure;
  - compatible `LayerCommunicator` instances;
  - available logits processor;
  - pipeline parallel size one.
- Capture completed post-residual hidden states using the existing layer-communication path.
- Preserve original forward signatures when runtime wrappers are attached.
- Return captured tensors as explicit transformer-body outputs so prefill CUDA-graph replay retains them.
- Resolve and forward runtime-captured states through logits processing.
- Gather auxiliary hidden states alongside normal hidden states in the context-parallel prefill path.
- Route Sarvam `TARGET_VERIFY` attention through:
  - the decode backend when `--speculative-attention-mode decode` is selected;
  - the prefill backend when `--speculative-attention-mode prefill` is selected.

### Scope and compatibility

The fallback is activated only when DSPARK auxiliary capture is requested. Normal inference and non-DSPARK models do not enter this path.

Models with native DSPARK or DFLASH capture hooks retain their existing behavior. Unsupported residual layouts and pipeline-parallel configurations fail during initialization instead of silently capturing incorrect tensors.

Regular Sarvam prefill and decode routing is unchanged. Only `TARGET_VERIFY` routing now follows the configured speculative attention mode.

## Accuracy Tests

Added focused unit coverage for:

- exact preservation of target output before and after runtime capture attachment;
- correct post-residual values at requested target layers;
- accumulator reset between forwards;
- native model-hook precedence;
- no-op behavior when auxiliary capture is disabled;
- invalid layer and model-contract rejection;
- logits handoff from ordinary and prefill-graph paths;
- no-op logits behavior without the runtime marker;
- Sarvam target-verification routing for both speculative attention modes.

The output-preservation test compares original and runtime-attached target outputs with `rtol=0` and `atol=0`.

The registered test passes through SGLang's direct CI-style runner:

```text
Ran 10 tests
OK
```

Focused and neighboring regression suites:

```text
29 passed, 8 parameterized subtests passed
```

A real-model smoke test was also completed on 2 x B200 with TP2 using the native `SarvamMLAForCausalLM` target, without a DeepSeek architecture override:

- DSPARK target and draft models loaded successfully.
- Torch compilation completed.
- Prefill CUDA graphs captured through 8192 tokens.
- Target and draft verification graphs captured for batch sizes 1-64.
- One 4,595-input-token request generated 64 tokens successfully.
- Eight concurrent requests with 2,025-6,524 input tokens each generated 64 tokens successfully.
- All requests returned HTTP 200.
- Server logs confirmed CUDA-graph replay for prefill and DSPARK decode.
- The concurrent smoke observed an accepted length of 3.35.

No task-level quality score is reported because this change does not modify target logits or model mathematics. DSPARK proposals remain target-verified, and exact preservation of the target output is directly tested.

## Speed Tests and Profiling

This PR provides runtime integration rather than a new performance kernel. There is no direct pre-patch native-Sarvam DSPARK comparison because the path could not be enabled without model-local capture support.

A representative throughput run was completed on 2 x B200, TP2, with approximately 4.65K input tokens and exactly 256 output tokens per request. The configuration used:

- TRT-LLM MLA decode;
- FA4 prefill;
- FlashInfer TRT-LLM BF16 MoE;
- torch compilation through request batch size 64;
- target and draft verification CUDA graphs through batch size 64;
- breakable prefill CUDA graphs through 8192 tokens;
- DSPARK block size 7.

The performance run used a reviewed DeepSeek-V3-compatible target implementation with Sarvam dimensions. That compatibility configuration is not part of this PR. The native Sarvam implementation was validated separately in the real-model smoke test above.

| Concurrency | Aggregate output tok/s | Decode tok/s/sequence | Accepted length |
|---:|---:|---:|---:|
| 1 | 317.78 | 367.6 | 2.989 |
| 8 | 820.57 | 111.5 | 2.905 |
| 32 | 1346.32 | 47.2 | 2.867 |
| 64 | 1576.25 | 29.8 | 2.944 |

All measured requests completed successfully. DSPARK acceptance remained approximately 27-28%, advancing approximately 2.9 output positions per target verification.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). N/A: this adds internal executor plumbing without introducing a public API, CLI flag or configuration field.
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :warning: [Run #35133548582](https://github.com/sgl-project/sglang/actions/runs/35133548582)
Latest PR Test (Extra): :warning: [Run #35133548148](https://github.com/sgl-project/sglang/actions/runs/35133548148)
Latest PR Test (AMD ROCm 10): :warning: [Run #35133548486](https://github.com/sgl-project/sglang/actions/runs/35133548486)

### CODE REVIEW

**Comment by tejesh-sarvam** [comment]
Hi @hnyls2002 @Qiaolin-Yu @Fridge003 , could you please review this DSPARK change and help trigger CI with /tag-and-rerun-ci?
